1 Introduction
---

In [18]:
%cat intro_gpu.log

0.9416
0.9636
0.9707
0.975
0.9771
0.9784
0.9757
0.9808
0.9815
0.9788
0.982
0.9808
0.9775
0.982
0.9809
0.982
0.9816
0.984
0.9836
0.9847
0.9848
0.9849
0.9849
0.9847
0.9847
0.9848
0.9848
0.9849
0.9848
0.9847
0.9847
0.9847
0.9847
0.9848
0.9848
0.9848
0.9847
0.9846
0.9846
0.9846
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9846
0.9846
0.9846
0.9846
0.9846
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9846
0.9846
0.9846
0.9846
0.9846
0.9846
0.9846
0.9846
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845
0.9845

real    3m33.077s
user    2m19.972s
sys     1m9.975s


2 Dropout
--

How does dropout work?
Each iteration a new subset of random activations is set to 0.
(Their gradient in the backwards pass is 0 as well)
This mainly breaks prevetns the codependence of features and forces the model to be more robust.
Often it can work as an effective form of regularization.

Why two models?
Dropout is used at training time, but not at prediction time.
Our implementation uses two different models for each task.

In [2]:
%cat ./dropout.log

Using gpu device 0: GeForce GTX 580 (CNMeM is disabled, CuDNN not available)
0.9367
0.9642
0.9733
0.9759
0.9762
0.9767
0.9785
0.9792
0.9803
0.9815
0.9838
0.9831
0.9829
0.9838
0.9825
0.9822
0.9833
0.9836
0.9834
0.9839
0.9846
0.9848
0.9835
0.9834
0.9841
0.9851
0.9854
0.9847
0.9848
0.9842
0.9853
0.9845
0.9842
0.9855
0.9846
0.9859
0.9845
0.985
0.985
0.9849
0.9855
0.9858
0.9858
0.9857
0.9858
0.9852
0.9859
0.9856
0.9858
0.9859
0.986
0.9867
0.9862
0.986
0.9862
0.9866
0.9859
0.986
0.9861
0.9857
0.9853
0.9872
0.9871
0.9875
0.9872
0.9859
0.9873
0.9876
0.9867
0.9868
0.9876
0.9871
0.987
0.9865
0.9873
0.987
0.9866
0.9863
0.9867
0.9863
0.9864
0.9864
0.987
0.9867
0.9876
0.9869
0.9864
0.9881
0.9874
0.9872
0.9871
0.9871
0.9862
0.9869
0.9861
0.9877
0.9879
0.9872
0.9875
0.9874

real	12m22.859s
user	23m30.126s
sys	0m45.546s


3 PReLU
--
Because we had problems with theano's broadcasting behaviour, we implemented the channel-shared prelu variant, also described in 'Delving Deep into Rectifiers'.

In [4]:
%cat prelu_layer-wise.log

correct: 0.9371, a: 0.11, a2: 0.94
correct: 0.9634, a: -0.03, a2: 0.96
correct: 0.9724, a: -0.08, a2: 0.97
correct: 0.9726, a: -0.11, a2: 0.97
correct: 0.9749, a: -0.13, a2: 0.97
correct: 0.9771, a: -0.15, a2: 0.98
correct: 0.9793, a: -0.16, a2: 0.98
correct: 0.9772, a: -0.18, a2: 0.98
correct: 0.9794, a: -0.18, a2: 0.98
correct: 0.9799, a: -0.19, a2: 0.98
correct: 0.9728, a: -0.19, a2: 0.97
correct: 0.9783, a: -0.18, a2: 0.98
correct: 0.9796, a: -0.18, a2: 0.98
correct: 0.9818, a: -0.18, a2: 0.98
correct: 0.9808, a: -0.17, a2: 0.98
correct: 0.9815, a: -0.18, a2: 0.98
correct: 0.9813, a: -0.18, a2: 0.98
correct: 0.9827, a: -0.18, a2: 0.98
correct: 0.9819, a: -0.18, a2: 0.98
correct: 0.9799, a: -0.18, a2: 0.98
correct: 0.9812, a: -0.18, a2: 0.98
correct: 0.9817, a: -0.17, a2: 0.98
correct: 0.9833, a: -0.18, a2: 0.98
correct: 0.9842, a: -0.17, a2: 0.98
correct: 0.9844, a: -0.17, a2: 0.98
correct: 0.9829, a: -0.17, a2: 0.98
correct: 0.9831, a: -0.18, a2: 0.98
correct: 0.9836, a: -0.19, a2

4 A simple ConvNet
---
4.1 Sketch
The colored bars highlight the convolutional layers.
It assuems border_mode='half', which is supposed to mimic the behavious or border_mode='same' from the scipy conv2d implementation. The default shrinks the image size by (k/2 -1), 'full' increases it by the same amount. 'half'/'same' keeps the size as before.

![title](architecture.png)

4.2 fancy plots
---

In [17]:
%cat convnet.log

In [7]: %time %run convnet.py
start
0.1225
0.4951
0.8969
0.9461
0.9644
0.9734
0.9772
0.9816
0.984
0.9876
0.9866
0.9882
0.9866
0.9889
0.9898
0.9887
0.9892
0.9901
0.9909
0.9911
CPU times: user 4min 37s, sys: 1min 55s, total: 6min 33s
Wall time: 6min 35s


The 20th datum of the test set
![title](test_20.png)


The three first filtermaps from the first conv layer
---
Note how the middle one resembles a vertical edge detection filter.
Interpretation of the other ones requires more imagination.
The right most one has a faint resemblance of a diagonal edeg detection filter.
![title](filters.png)

The filtermaps applied to the test datum
---
The right most filter actually highlights diagonal edges!
![title](filtered.png)

4.3 extra task
---
Since all the models show good performance on the mnist data, it is not very interesting to increase the capability of the convnet.
Instead, we want to see how it deteriorates. This gives us two possibilites, either increase the size until we run into convergence problems, or reduce the numbers of layers.
For the sake of saving no computational and student resources, we chose the latter and removed the third convolutional layer.

Of course it is a bad idea to just flatten the old output of the second layer, because it has 64x5x5 weights
Instead the secodn layer now has only 16 feature maps, which gives a fullyconnected layer with 265 nodes. The hidden representation has only half the size of our prior architecture.

In [19]:
%cat convnet_small.log

In [23]: %time %run convnet_small.py
start
0.846
0.9495
0.9698
0.975
0.9809
0.9829
0.9857
0.987
0.987
0.9875
0.9881
0.9882
0.9889
0.9901
0.9901
0.9908
0.9904
0.9907
0.991
0.9902
CPU times: user 3min 17s, sys: 1min 26s, total: 4min 44s
Wall time: 4min 45s


5 Test Error Comparison
---


architecture | classification rate
--|--
basic | 0.9845
dropout| 0.9874
prelu | 0.9839
convnet | 0.9911
extra convnet | 0.9902

All models perform well.
The non-convolutional were all trained for five times as many epochs as the convolutional ones.
But all models converged in clearly less than the 